### PR TITLE
HyperNova Gadgets

### PR DESCRIPTION
**Pull request by sjudson**
Working towards: https://github.com/nexus-xyz/nexus-zkvm/issues/76

This PR implements HyperNova (CycleFold) folding gadgets for constraints derived from R1CS.

### CODE REVIEW

File: nova/src/gadgets/cyclefold/hypernova/primary.rs [hunk 1]
```diff
@@ -0,0 +1,561 @@
...
+        Self {
+            commitment_W: self.commitment_W.clone(),
+            X: self.X.clone(),
+            _commitment_scheme: self._commitment_scheme,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct CCSInstanceFromR1CSVar<G1, C1>(CCSInstanceVar<G1, C1>)
```
Comment (sjudson) [inline nova/src/gadgets/cyclefold/hypernova/primary.rs:52]: I wanted to make the implementation explicit about the necessary source of the constraints, as well as to interface with `ark-r1cs-std::R1CSVar` in a clear way, and this seemed like the most minimal way to do so. Happy to consider other approaches though.

File: nova/src/ccs/mod.rs [hunk 2]
```diff
@@ -302,6 +331,17 @@ impl<G: CurveGroup, C: PolyCommitmentScheme<G>> LCCSInstance<G, C> {
         }
     }
 
+    /// A method to create a default instance object with an appropriate shape for the base case of IVC.
+    pub fn base(shape: &CCSShape<G>) -> Self {
```
Comment (slumber) [inline nova/src/ccs/mod.rs:335]: I don't think you will have `shape` available in the base case of the ivc, unless you either introduce lifetimes everywhere and borrow it.

Seems irrelevant for this PR, but also insignificant.

**Review by slumber — APPROVED** [review]
Looks great!

Absorbing LCCS instance with a huge step circuit might increase the recursion overhead (`rs.len()` grows) by several thousands, but the multifold itself has 30% less constraints than original Nova.

File: nova/src/ccs/mod.rs (hunk 2 above)
Comment (sjudson) [reply nova/src/ccs/mod.rs:335]: Ah this was a part of sketching out the IVC and I forgot to cut it from this PR along with the rest. I'll remove it.
